### ISSUE TITLE
Problems with scene graph and debug renderer

### ISSUE THREAD

**Issue by wivlaro** [mosra/magnum#79]
I've been trying to figure out a problem with the scene graph I've been having whereby I get complaints about the rotation 3x3 part of child objects' transformation matrix not being evenly scaled. I have tried to make a simple program that uses parenting, but I'm having trouble displaying anything at all.

Now I have tried to use the debug renderers as documented here http://mosra.cz/blog/magnum-doc/debug-tools.html#debug-tools-renderers but there are some problems with the resource manager not being correctly initialized.

For this example, I have the dreaded blank screen anyway, the simpledraw functions are just trying to make SOMETHING appear.

If anyone can shed any light on either the resource manager problem, or why the scene graph doesn't seem to be combining matrices properly, it would be appreciated.

Complete code (stack trace below):
```c++
#include <vector>
#include <iostream>
#include <memory>

#include <Magnum/DefaultFramebuffer.h>
#include <Magnum/Platform/Sdl2Application.h>
#include <Magnum/SceneGraph/Scene.h>
#include <Magnum/SceneGraph/MatrixTransformation3D.h>
#include <Magnum/SceneGraph/AbstractCamera.h>
#include <Magnum/SceneGraph/Camera3D.h>
#include <Magnum/SceneGraph/Drawable.h>
#include <Magnum/Buffer.h>
#include <Magnum/Math/Matrix4.h>
#include <Magnum/Shaders/VertexColor.h>
#include <Magnum/Mesh.h>
#include <Magnum/DebugTools/ResourceManager.h>
#include <Magnum/DebugTools/ObjectRenderer.h>
#include <Magnum/Renderer.h>
#include <Corrade/PluginManager/AbstractManager.h>

using namespace Magnum;

typedef SceneGraph::Object<SceneGraph::MatrixTransformation3D> Object3D;
typedef SceneGraph::Scene<SceneGraph::MatrixTransformation3D> Scene3D;

class MyObject3D;

class MyApplication: public Platform::Application {
    public:
        explicit MyApplication(const Arguments& arguments);

    private:
        void drawEvent() override;
		void mouseMoveEvent(MouseMoveEvent& event) override;

        Scene3D scene;
        Object3D* cameraObject;
        SceneGraph::Camera3D* camera;
        SceneGraph::DrawableGroup3D drawables;
		
		MyObject3D* obj;
		MyObject3D* obj2;
		
		// Global instance of debug resource manager, drawable group for the renderers
		DebugTools::ResourceManager manager;
		SceneGraph::DrawableGroup3D debugDrawables;
};

class MyObject3D : public Object3D, SceneGraph::Drawable3D {
public:
	MyObject3D(const std::vector<Vector3>& data, Object3D* parent, SceneGraph::DrawableGroup3D* drawables)
	:	Object3D(parent)
	,	SceneGraph::Drawable3D(*this, drawables)
	{
		_buffer.setData(data, BufferUsage::StaticDraw);
		_mesh.setPrimitive(MeshPrimitive::Triangles)
			.setCount(data.size() / 2)
			.addVertexBuffer(_buffer, 0, Shaders::VertexColor3D::Position{}, Shaders::VertexColor3D::Color{});
	}
    void simpledraw() {
		_mesh.draw(_shader);
	}
	
private:
    void draw(const Matrix4& transformationMatrix, SceneGraph::AbstractCamera3D& camera) override {
		_shader.setTransformationProjectionMatrix(transformationMatrix);
		_mesh.draw(_shader);
	}
	Buffer _buffer;
	Mesh _mesh;
	Shaders::VertexColor3D _shader;
};


MyApplication::MyApplication(const Arguments& arguments): Platform::Application(arguments) {
	CORRADE_RESOURCE_INITIALIZE(MagnumShaders_RCS);
    /* Configure camera */
    cameraObject = new Object3D(&scene);
    cameraObject->translate(Vector3::zAxis(5.0f));
    camera = new SceneGraph::Camera3D(*cameraObject);
    camera->setAspectRatioPolicy(SceneGraph::AspectRatioPolicy::Extend)
        .setPerspective(35.0_degf, 4.0f/3.0f, 0.001f, 1000.0f)
        .setViewport(defaultFramebuffer.viewport().size());
		
	std::vector<Vector3> data {
		{-0.5f, -0.5f, 0.0f}, {1.0f, 0.0f, 0.0f}, /* Left vertex, red color */
		{ 0.5f, -0.5f, 0.0f}, {0.0f, 1.0f, 0.0f}, /* Right vertex, green color */
		{ 0.0f,  0.5f, 0.0f}, {0.0f, 0.0f, 1.0f}  /* Top vertex, blue color */
	};
	std::vector<Vector3> data2 = {
		{-0.5f, -0.5f, 0.0f}, {1.0f, 1.0f, 0.0f}, /* Left vertex, red color */
		{ 0.5f, -0.5f, 0.0f}, {0.0f, 1.0f, 1.0f}, /* Right vertex, green color */
		{ 0.0f,  0.5f, 0.0f}, {1.0f, 0.0f, 1.0f}  /* Top vertex, blue color */
	};
	
		
	auto debugRendererOptions = new DebugTools::ObjectRendererOptions();
	debugRendererOptions->setSize(0.3f);
	
	obj = new MyObject3D(data, &scene, &drawables);
	obj->setTransformation(Matrix4::translation({0.1f,0.2f,0.3f}) * Matrix4::rotation(0.2_degf, Vector3{1.0,0.0,0.0}));
	obj2 = new MyObject3D(data2, obj, &drawables);
	obj2->setTransformation(Matrix4::translation({0.1f,0.2f,0.3f}) * Matrix4::rotation(-0.1_degf, Vector3{-0.41f, 0.0f, 0.41f}.normalized()));
	
	DebugTools::ResourceManager::instance().set("my", DebugTools::ObjectRendererOptions().setSize(0.3f));
	new DebugTools::ObjectRenderer3D(*obj, "my", &debugDrawables);
	new DebugTools::ObjectRenderer3D(*obj2, "my", &debugDrawables);
}

void MyApplication::mouseMoveEvent(MouseMoveEvent& event) {
//     if(!(event.buttons() & MouseMoveEvent::Button::Left)) return;

	auto xrot = Rad(event.position().x() / 100.0f);
	auto yrot = Rad(event.position().y() / 100.0f);
	std::cout << "xrot " << Float(Deg(xrot)) << " yrot " << Float(Deg(yrot)) << std::endl;
	Matrix4 cameraTransformation = 
		Matrix4::rotationY(xrot)
		*
		Matrix4::translation({0,0,50})
// 		*
// 		Matrix4::rotationZ(yrot)
	;
	
	
	
	cameraObject->setTransformation(
		cameraTransformation
	);
	
    event.setAccepted();
    redraw();
}

void MyApplication::drawEvent() {
    defaultFramebuffer.clear(FramebufferClear::Color|FramebufferClear::Depth);

	Renderer::disable(Renderer::Feature::DepthTest);
	Renderer::enable(Renderer::Feature::DebugOutput);
    camera->draw(drawables);
    camera->draw(debugDrawables);
	
	obj->simpledraw();
	obj2->simpledraw();
	

    swapBuffers();
}

MAGNUM_APPLICATION_MAIN(MyApplication)
```

Here is the stack trace, it looks like the std::map is not properly initialized. I've been using resource manager before in other places and haven't had problems.
```
#0 std::_Hashtable<Magnum::ResourceKey, std::pair<Magnum::ResourceKey const, Magnum::Implementation::ResourceManagerData<Magnum::DebugTools::ObjectRendererOptions>::Data>, std::allocator<std::pair<Magnum::ResourceKey const, Magnum::Implementation::ResourceManagerData<Magnum::DebugTools::ObjectRendererOptions>::Data> >, std::__detail::_Select1st, std::equal_to<Magnum::ResourceKey>, std::hash<Magnum::ResourceKey>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::_M_find_before_node() at /usr/include/c++/4.9/bits/hashtable.h:1442
#1 std::_Hashtable<Magnum::ResourceKey, std::pair<Magnum::ResourceKey const, Magnum::Implementation::ResourceManagerData<Magnum::DebugTools::ObjectRendererOptions>::Data>, std::allocator<std::pair<Magnum::ResourceKey const, Magnum::Implementation::ResourceManagerData<Magnum::DebugTools::ObjectRendererOptions>::Data> >, std::__detail::_Select1st, std::equal_to<Magnum::ResourceKey>, std::hash<Magnum::ResourceKey>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::_M_find_node() at /usr/include/c++/4.9/bits/hashtable.h:625
#2 std::_Hashtable<Magnum::ResourceKey, std::pair<Magnum::ResourceKey const, Magnum::Implementation::ResourceManagerData<Magnum::DebugTools::ObjectRendererOptions>::Data>, std::allocator<std::pair<Magnum::ResourceKey const, Magnum::Implementation::ResourceManagerData<Magnum::DebugTools::ObjectRendererOptions>::Data> >, std::__detail::_Select1st, std::equal_to<Magnum::ResourceKey>, std::hash<Magnum::ResourceKey>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::find() at /usr/include/c++/4.9/bits/hashtable.h:1304
#3 std::unordered_map<Magnum::ResourceKey, Magnum::Implementation::ResourceManagerData<Magnum::DebugTools::ObjectRendererOptions>::Data, std::hash<Magnum::ResourceKey>, std::equal_to<Magnum::ResourceKey>, std::allocator<std::pair<Magnum::ResourceKey const, Magnum::Implementation::ResourceManagerData<Magnum::DebugTools::ObjectRendererOptions>::Data> > >::find() at /usr/include/c++/4.9/bits/unordered_map.h:574
#4 Magnum::Implementation::ResourceManagerData<Magnum::DebugTools::ObjectRendererOptions>::set() at /usr/local/include/Magnum/ResourceManager.h:507
#5 Magnum::ResourceManager<Magnum::AbstractShaderProgram, Magnum::Buffer, Magnum::Mesh, Magnum::MeshView, Magnum::DebugTools::ForceRendererOptions, Magnum::DebugTools::ObjectRendererOptions, Magnum::DebugTools::ShapeRendererOptions>::set<Magnum::DebugTools::ObjectRendererOptions>() at /usr/local/include/Magnum/ResourceManager.h:313
#6 Magnum::ResourceManager<Magnum::AbstractShaderProgram, Magnum::Buffer, Magnum::Mesh, Magnum::MeshView, Magnum::DebugTools::ForceRendererOptions, Magnum::DebugTools::ObjectRendererOptions, Magnum::DebugTools::ShapeRendererOptions>::set<Magnum::DebugTools::ObjectRendererOptions>() at /usr/local/include/Magnum/ResourceManager.h:330
#7 Magnum::ResourceManager<Magnum::AbstractShaderProgram, Magnum::Buffer, Magnum::Mesh, Magnum::MeshView, Magnum::DebugTools::ForceRendererOptions, Magnum::DebugTools::ObjectRendererOptions, Magnum::DebugTools::ShapeRendererOptions>::set<Magnum::DebugTools::ObjectRendererOptions&>() at /usr/local/include/Magnum/ResourceManager.h:335
#8 MyApplication::MyApplication() at /home/bill/programming/magnum-scenegraph3D/src/MyApplication.cpp:105
#9 main() at /home/bill/programming/magnum-scenegraph3D/src/MyApplication.cpp:149
```
Also, I have all the libraries and plugins set to statically link, if that makes any difference.

**Comment by wivlaro** [comment]
Thanks very much for your help with the example. I figured out my rotation matrix problem in the end. I was reading it in from an external source and it needed transposing.

**Comment by mosra** [comment]
Let's keep it open, I still need to resolve the issue with static build and make the documentation suck less :)

**Comment by mosra** [comment]
Documentation should be hopefully improved with 4caeb30f7b996a19b3a460d472c387be9eadadc9 and 9fd25a1fd8fa38ae1f761ddd69b29b183ab12bf2 (also already uploaded to [web](http://mosra.cz/blog/magnum-doc/)), now focusing on the static build ...

**Comment by mosra** [comment]
Finally got time to look into the static build, but couldn't reproduce (static build of Corrade + Magnum) with the code you posted above. Fixed possibly related issue in 7e8d90a436fe57ee236ea854ef69cab828b25118 and added additional test to verify that this kind of problems won't appear again.

If you come across this crash again, please reopen.
